### PR TITLE
Support validation contexts as Array

### PR DESCRIPTION
**Pull request by McRip**
Hello Justin,

i added support for displaying a field as required, when the validation context in the model is specified as an array to have multiple validation contexts for a validation.
I would be very happy if this could be merged.

Thanks in advance
Moritz

### CODE REVIEW

File: lib/formtastic/inputs/base/validations.rb [hunk 1]
```diff
@@ -138,7 +138,11 @@ def required?
           if validations?
             validations.select { |validator|
               if validator.options.key?(:on)
-                return false if (validator.options[:on] != :save) && ((object.new_record? && validator.options[:on] != :create) || (!object.new_record? && validator.options[:on] != :update))
+                if validator.options[:on].is_a?(Array)
+                  return false if (validator.options[:on].exclude?(:save)) && ((object.new_record? && validator.options[:on].exclude?(:create)) || (!object.new_record? && validator.options[:on].exclude?(:update)))
```
Comment (mikz) [inline lib/formtastic/inputs/base/validations.rb]: Could this be simplified by always converting it to an Array like `Array(validator.options[:on])` and then just checking for `exclude?` ?
Having similar conditions in two branches is really dangerous.

Comment (st0012) [inline lib/formtastic/inputs/base/validations.rb]: I think this can be more simple because you don't actually need that `if` statement, just return the opposite value will do the job
```ruby
!(validator.options[:on].exclude?(:save) &&
  ((object.new_record? && validator.options[:on].exclude?(:create)) ||
   (!object.new_record? && validator.options[:on].exclude?(:update))
))
```
And you can also create an private method so that the condition can be more readable
```ruby
def exclude_action?(action)
  validator.options[:on].exclude?(action)
end
```

```ruby
!(exclude_action?(:save) &&
  ((object.new_record? && exclude_action?(:create)) ||
   (!object.new_record? && exclude_action?(:update))
))
```

Just some advise :smile:

**Comment by McRip** [comment]
Thank you for your comments. You were right, i was a bit quick on this. I prefer @mikz solution to just convert it to an array and then have only one check left. I changed my commit.

File: lib/formtastic/inputs/base/validations.rb [hunk 2]
```diff
@@ -138,7 +138,8 @@ def required?
           if validations?
             validations.select { |validator|
               if validator.options.key?(:on)
-                return false if (validator.options[:on] != :save) && ((object.new_record? && validator.options[:on] != :create) || (!object.new_record? && validator.options[:on] != :update))
+                validator_on = validator.options[:on].is_a?(Array) ? validator.options[:on] : [validator.options[:on]]
```
Comment (mikz) [inline lib/formtastic/inputs/base/validations.rb]:
```ruby
validator.options[:on].is_a?(Array) ? validator.options[:on] : [validator.options[:on]]
```
is the same as
```ruby
Array(validator.options[:on])
```
`Array` is also a function. See http://docs.ruby-lang.org/en/2.2.0/Kernel.html#method-i-Array.

**Comment by McRip** [comment]
Thank you @mikz, you are so right. I changed my implementation and pushed a new version of my PR.
@justinfrench: Can this be merged?

**Comment by mikz** [comment]
@McRip would it be possible to have a real unit test for that logic? Because the tests you added cover the case when you add an array, but does not test what happens when you have there both `:save` and `:create` or different combinations. Testing this on the 'integration' level is good for some examples, but the edge cases could be tested on lower level which would reduce amount of tests and noise they add. But that is just my 2c. Just a nitpick.

**Comment by McRip** [comment]
@mikz I created unit tests for Formtastic::Inputs::Base::Validations#required?. I hope this can be merged soon, so that i can ditch my git dependency in my Gemfile :)

**Comment by justinfrench** [comment]
@mikz Looks good to me. I'll merge in a couple of days, but hold off in case you have additional comments.

**Comment by mikz** [comment]
@McRip wow! Very nice :+1: 
@justinfrench looks great to me!

**Comment by justinfrench** [comment]
Thanks @McRip!

**Comment by McRip** [comment]
Thanks for merging!
